pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/MultipleOrEqualitiesToInClauseFilterQueryTreeOptimizer.java

```diff
@@ -20,148 +20,291 @@
     for (FilterQueryTree childQueryTree : filterQueryTree.getChildren()) {
-      if (childQueryTree.getOperator() == FilterOperator.EQUALITY
-          || childQueryTree.getOperator() == FilterOperator.IN) {
+      FilterOperator operator = childQueryTree.getOperator();
+      if (operator == FilterOperator.EQUALITY || operator == FilterOperator.IN) {
+        // the immediate child of root OR operator is EQUALITY or IN
         List<String> childValues = childQueryTree.getValue();
-        if (!columnToValues.containsKey(childQueryTree.getColumn())) {
-          TreeSet<String> value = new TreeSet<>(childValues);
-          columnToValues.put(childQueryTree.getColumn(), value);
-          if (!containsDuplicates && value.size() != childValues.size()) {
-            containsDuplicates = true;
+        String column = childQueryTree.getColumn();
+        Set<String> predicateValues = columnToInEqPredicateValues.get(column);
+        if (predicateValues == null) {
+          // we didn't see this column before
+          predicateValues = new HashSet<>(childValues);
+          // if after adding to the set, the number of predicate values become different (less),
+          // it implies that values got deduplicated and so we need to rewrite this predicate later
+	  // similarly, if this is a single value IN predicate, we need to rewrite it later to equality
+          int numChildren = childValues.size();
+          if (operator == FilterOperator.IN && (numChildren == 1 || numChildren != predicateValues.size())) {
+            rewriteINEQChildrenOfORRootOperator = true;
           }
+          columnToInEqPredicateValues.put(column, predicateValues);
         } else {
-          Set<String> currentValues = columnToValues.get(childQueryTree.getColumn());
-          for (String childValue : childValues) {
-            if (!containsDuplicates && currentValues.contains(childValue)) {
-              containsDuplicates = true;
-            } else {
-              currentValues.add(childValue);
-            }
-          }
+          // if we had earlier seen this column, we need to rewrite the predicate
+          predicateValues.addAll(childValues);
+          rewriteINEQChildrenOfORRootOperator = true;
         }
       } else {
-        nonEqualityOperators.add(childQueryTree);
-      }
-    }
-
-    return containsDuplicates;
-  }
-
-  private boolean isRebuildRequired(Map<String, Set<String>> columnToValues, boolean containsDuplicates) {
-    // We need to rebuild the predicate if there were duplicate values detected (eg. a = 1 OR a = 1) or if there is
-    // more than one value for a column (eg. a = 1 OR a = 2)
-    boolean rebuildRequired = containsDuplicates;
-
-    if (!rebuildRequired) {
-      for (Set<String> columnValues : columnToValues.values()) {
-        if (1 < columnValues.size()) {
-          rebuildRequired = true;
-          break;
+        // the immediate child of root operator is <, >, AND, RANGE, <=, >=, BETWEEN ...
```

Comment:
(nit) all the comparisons are classified as RANGE

Reply:
Done